### DISCUSSION TITLE
Move-on from Ksonnet library

### DISCUSSION THREAD

**Discussion by carlosedp** [carlosedp/cluster-monitoring#121] [General]
We need a plan to move-on from the deprecated Ksonnet JSONnet library to something more up-to-date.

I'm open to suggestions but won't be totally available right now to do it all by myself.

Some options I see:

* Pulumi - https://github.com/pulumi/pulumi
* Grafana Tanka - https://github.com/grafana/tanka
* Red Hat Openshift libraries - https://github.com/openshift/cluster-monitoring-operator/tree/master/jsonnet

**Comment by Kolossi** [comment]
As another option, why not just go back to base jsonnet as kube-prometheus itself has done?

**Comment by ToMe25** [comment]
Cian911 suggested using https://github.com/jsonnet-libs/k8s-libsonnet [here](https://github.com/carlosedp/cluster-monitoring/issues/91#issuecomment-939053344), which would probably be comparatively easy to use as a replacement.

I personally however think that using base jsonnet would likely be the best choice, tho this is partly simply because the syntax seems easier to read for me.

**Comment by cyc1ingsir** [comment]
I just tested how much effort would be needed to port this to jsonnet-libs/k8s-libsonnet. At the first glance, it seems to be quite straight forward.(1)   
However, I noticed that this library is providing branches/releases coupled with K8s versions. The upstream kube-prometheus project has a compatibility matrix stating the compatibility of two K8s versions for every release. Wouldn't something similar be needed for this project to cover a range of K8s (K3s) versions?

(1) Without enabling any extras and minimizing the usage of addons, I managed to generate manifests out of this
https://github.com/cyc1ingsir/cluster-monitoring/commit/5f6d0368e04938744aa30b5237963eae25ae8a68
by `make update_libs; make vendor; make`
Havn't had the chance to test the deployment of those.

**Reply by cyc1ingsir** [reply]
Slowly I'm getting there:
Using 
 - K3s (version: v1.22.3+k3s1)
 - kube-prometheus (version: v0.9.0)
 - jsonnet-libs/k8s-libsonnet (version: 1.22)
I'm able to generate most of the manifests again and deploy them on the test cluster running on a single RaspPi 4B with arm64 Raspi OS.

It's far from being complete but anyone interested could have a look here:
https://github.com/cyc1ingsir/cluster-monitoring/tree/feature/porting_to_jsonnet-libs_k8s-libsonnet
